File: skills/hypatia/scripts/render.py
"""Markdown, SVG, Beamer, and provenance exports from validated evidence."""

import re
from uuid import uuid4

from evidence import evidence_digest, evidence_directory, load_context, load_evidence, validate_delivery
from handoff import load_snapshot
from slides import beamer, opportunity_text, shortlist
from storage import atomic_write, digest, inside, json_text


def md(value: str) -> str:
    return re.sub(r"\r?\n", " ", re.sub(r"[\\`*_{}\[\]()<>#|!]", r"\\\g<0>", value))


def xml(value: str) -> str:
    return value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace('"', "&quot;")


def cites(ids):
    return ", ".join(f"[{claim_id}](#claim-{claim_id})" for claim_id in ids)


def report(snapshot, evidence, context):
    handoff, ledger = snapshot["handoff"], snapshot["ledger"]
    lines = [
        f"# Hypatia — {md(ledger['question'])}", "",
        f"Audience: {md(context['audience'])}. Search cutoff: **{handoff['cutoff']}**.",
        f"Callimachus revision: `{handoff['revision']}`. Final included publications: **{len(handoff['included_ids'])}**.",
        "Publication counts are not independent-study counts or evidence-quality scores.", "",
        "## Executive summary", "",
        *(f"- {md(f['statement'])} ({cites(f['claim_ids'])})" for f in evidence["findings"][:5]),
        "" if evidence["findings"] else "No supported findings were established.",
        f"{len(shortlist(evidence))} direction(s) meet the stated low-effort criteria. Feasibility is Hypatia's assessment, not an author claim.",
        "", "## Scope and methods", "",
        f"Include: {'; '.join(map(md, ledger['criteria']['include'])) or 'None specified'}.",
        f"Exclude: {'; '.join(map(md, ledger['criteria']['exclude'])) or 'None specified'}.",
        f"Callimachus recorded {len(ledger['queries'])} query audit entries and {len(ledger['records'])} deduplicated records. All abstract screening and required human curation dispositions are complete.",
        "The complete search audit and screening decisions are in the pinned Callimachus ledger. No additional search was performed by Hypatia.",
        "", "## Main findings and disagreements", "",
    ]
    for finding in evidence["findings"]:
        lines.extend([
            f"### {finding['id']}: {md(finding['statement'])}", "",
            f"Evidence: {cites(finding['claim_ids'])}. Appraisal: {md(finding['strength'])}.",
            f"Disagreement / counterevidence: {cites(finding['disagreements']) if finding['disagreements'] else 'none recorded; this does not establish consensus'}.", "",
        ])
    lines.extend(["## Literature-identified gaps", ""])
    for gap in evidence["gaps"]:
        lines.extend([
            f"### {gap['id']}: {md(gap['statement'])}", "",
            f"Status: **{gap['status']}**. Evidence: {cites(gap['claim_ids'])}.",
            f"Currency assessment: {md(gap['currency']['rationale'])} ({len(gap['currency']['checked_source_ids'])}/{len(handoff['included_ids'])} included publications checked).",
            f"Subsequent evidence: {cites(gap['currency']['claim_ids']) or 'none recorded'}.", "",
        ])
    lines.extend([
        "An open gap is scoped to this completed corpus and its cutoff; it is not proof of absence across the field.", "",
        "## Low-hanging fruit under stated constraints", "",
    ])
    if not shortlist(evidence):
        lines.extend(["No directions can currently be shortlisted with adequate evidence and known resources.", ""])
    for opportunity in shortlist(evidence):
        lines.extend([
            f"- **{opportunity['id']}** — {md(opportunity_text(evidence, opportunity))} ({cites(opportunity['direction_claim_ids'])}); see the assessment below.", "",
        ])
    lines.extend(["## All author-proposed directions and feasibility", ""])
    for opportunity in evidence["opportunities"]:
        feasibility = opportunity["feasibility"]
        lines.extend([
            f"### {opportunity['id']}: {md(opportunity_text(evidence, opportunity))}", "",
            f"Author-proposal evidence: {cites(opportunity['direction_claim_ids'])}. Gap: {opportunity['gap_id']}.",
            f"Hypatia effort assessment: **{feasibility['effort']}** — {md(feasibility['rationale'])}.",
            f"Prerequisites: {'; '.join(map(md, feasibility['prerequisites'])) or 'unknown'}.",
            f"Unknowns: {'; '.join(map(md, feasibility['unknowns'])) or 'none recorded'}.",
            f"Researcher resource references: {', '.join(feasibility['resource_ids']) or 'none; feasibility remains unestablished'}.", "",
        ])
    lines.extend([
        "## Researcher constraints", "",
        *(f"- {r['id']}: {md(r['description'])}" for r in context["resources"]),
        "" if context["resources"] else "No resources supplied; low-effort shortlisting is disabled.", "",
        "## Coverage, access, and unresolved evidence", "",
    ])
    for source in handoff["sources"]:
        limitation = f" — {md(source['limitation'])}" if source["limitation"] else ""
        lines.append(f"- {md(source['record_id'])}: {source['kind']}{limitation}. {' '.join(map(md, source['warnings']))}")
    for review in evidence["source_reviews"]:
        lines.append(f"- {md(review['source_id'])} — {review['status']}: {md(review['note'])}")
    for claim in evidence["claims"]:
        if claim["verification"]["status"] != "supported":
            lines.append(f"- {claim['id']}: **{claim['verification']['status']}** — {md(claim['verification']['reason'])}.")
    lines.extend(["", "## References", ""])
    for record in ledger["records"]:
        if record["id"] in handoff["included_ids"]:
            year = record["year"] if record["year"] is not None else "year unknown"
            lines.append(f"- **{md(record['id'])}**: {md(', '.join(record['authors']))}. {md(record['title'])} ({year}). DOI: {md(record['doi'] if record['doi'] is not None else 'unavailable')}.")
    lines.extend([
        "", "## Claim-to-source audit", "",
        "Exact quotation matching checks provenance. Semantic support and scientific strength remain explicit, reviewable assessments.", "",
    ])
    for claim in evidence["claims"]:
        source = next(s for s in handoff["sources"] if s["record_id"] == claim["source_id"])
        locator = "abstract" if source["kind"] == "abstract" else f"PDF page {claim['page']}"
        lines.extend([
            f'<a id="claim-{claim["id"]}"></a>', f"### {claim['id']} — {claim['kind']}", "",
            md(claim["statement"]), "", f"> {md(claim['quote'])}", "",
            f"Source: {md(claim['source_id'])}, {locator}; artifact `{source['artifact']}`.",
            f"Verification: **{claim['verification']['status']}** — {md(claim['verification']['reason'])}.",
            f"Appraisal: {md(claim['appraisal'])}. Independent-study identity: {md(claim['study_id'] if claim['study_id'] is not None else 'unknown')}.", "",
        ])
    return "\n".join(lines) + "\n"


def wrap(value, width=110):
    return [s.strip() for s in re.findall(rf".{{1,{width}}}(?:\s|$)|.{{1,{width}}}", value)] or [""]


def svg(title, description, rows):
    lines = [title, description]
    for row in rows:
        lines.extend([*wrap(row), ""])
    content = "\n".join(
        f'<text x="24" y="{38 + i * 25}" font-size="{22 if i == 0 else 15}">{xml(line)}</text>'
        for i, line in enumerate(lines)
    )
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" role="img" aria-labelledby="title description" viewBox="0 0 1150 {70 + len(lines) * 25}">\n'
        f'<title id="title">{xml(title)}</title><desc id="description">{xml(description)}</desc>\n'
        f'<rect width="100%" height="100%" fill="white"/><g fill="#172b4d" font-family="sans-serif">{content}</g></svg>\n'
    )


def table_svg(title, description, columns, rows):
    elements = []
    y = 125
    for index, row in enumerate([[c["label"] for c in columns], *rows]):
        cells = [wrap(value, max(8, int((columns[i]["width"] - 24) / 8.5))) for i, value in enumerate(row)]
        height = max(map(len, cells)) * 23 + 24
        x = 25
        for i, lines in enumerate(cells):
            fill = "#e6edf5" if index == 0 else "#ffffff" if index % 2 else "#f4f7fb"
            elements.append(f'<rect x="{x}" y="{y}" width="{columns[i]["width"]}" height="{height}" fill="{fill}" stroke="#bdcbdc"/>')
            elements.extend(
                f'<text x="{x + 12}" y="{y + 26 + line * 23}" font-weight="{600 if index == 0 else 400}">{xml(value)}</text>'
                for line, value in enumerate(lines)
            )
            x += columns[i]["width"]
        y += height
    caption = "\n".join(f'<text x="25" y="{67 + i * 22}">{xml(line)}</text>' for i, line in enumerate(wrap(description, 115)))
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" role="img" aria-labelledby="title description" viewBox="0 0 1150 {y + 35}">\n'
        f'<title id="title">{xml(title)}</title><desc id="description">{xml(description)}</desc>\n'
        '<rect width="100%" height="100%" fill="white"/>\n'
        '<g fill="#172b4d" font-family="sans-serif" font-size="15">\n'
        f'<text x="25" y="38" font-size="24" font-weight="600">{xml(title)}</text>\n'
        f'{caption}\n' + "\n".join(elements) + "</g></svg>\n"
    )


def render(snapshot):
    load_snapshot(snapshot["root"], snapshot["handoff"]["revision"])
    evidence, context = load_evidence(snapshot), load_context(snapshot)
    validate_delivery(snapshot, evidence)
    evidence_hash = evidence_digest(snapshot)
    renderer_version = 3
    export_hash = digest(json_text({"evidence_digest": evidence_hash, "renderer_version": renderer_version}))
    root = evidence_directory(snapshot)
    directory = inside(root, f"exports/{export_hash}")
    delivery = json_text({"digest": export_hash, "directory": str(directory)})
    if directory.exists():
        atomic_write(inside(root, "delivery.json"), delivery)
        return str(directory)
    staging = inside(root, f"exports/{export_hash}.{uuid4()}.partial")
    staging.mkdir(parents=True)

    def save(name, data):
        with (staging / name).open("xb") as stream:
            stream.write(data.encode("utf-8") if isinstance(data, str) else data)

    save("report.md", report(snapshot, evidence, context))
    save("slides.tex", beamer(snapshot, evidence, context))
    save("brief.md", "\n".join([
        f"# {md(snapshot['ledger']['question'])}",
        f"Search cutoff: {snapshot['handoff']['cutoff']}; revision: {snapshot['handoff']['revision']}.",
        "", "## Findings",
        *(f"- {md(f['statement'])} ({', '.join(f['claim_ids'])}; see report audit). Strength: {md(f['strength'])}. Counterevidence: {', '.join(f['disagreements']) or 'none recorded'}." for f in evidence["findings"]),
        "", "## Gaps",
        *(f"- {md(g['statement'])} — {g['status']} ({', '.join(g['claim_ids'])})." for g in evidence["gaps"]),
        "", "## Feasible author-proposed directions",
        *(f"- {md(opportunity_text(evidence, o))} ({', '.join(o['direction_claim_ids'])}). Hypatia feasibility assessment: {md(o['feasibility']['rationale'])}." for o in shortlist(evidence)),
        "" if shortlist(evidence) else "No adequately supported low-effort shortlist.",
        "", "## Limitations", "This summary inherits the report's corpus cutoff, access limitations, uncertainty, and researcher constraints.",
        *(f"- {md(s['record_id'])}: {md(s['limitation'])}" for s in snapshot["handoff"]["sources"] if s["limitation"]),
        *(f"- {c['id']}: {c['verification']['status']} — {md(c['verification']['reason'])}" for c in evidence["claims"] if c["verification"]["status"] != "supported"),
    ]) + "\n")
    caption = f"Cutoff {snapshot['handoff']['cutoff']}. {len(snapshot['handoff']['included_ids'])} included publications; counts are not evidence strength."

    def cells(source_id, kinds):
        return "; ".join(
            f"{c['id']} ({c['verification']['status']})"
            for c in evidence["claims"] if c["source_id"] == source_id and c["kind"] in kinds
        ) or "No extracted claims"

    save("evidence-matrix.svg", table_svg("Evidence matrix — claim IDs and verification status", caption, [
        {"label": "Publication / access", "width": 240}, {"label": "Findings", "width": 215},
        {"label": "Gaps / limitations", "width": 215}, {"label": "Author proposals", "width": 215},
        {"label": "Context", "width": 215},
    ], [
        [f"{s['record_id']} [{s['kind']}]", cells(s["record_id"], ["finding"]), cells(s["record_id"], ["gap", "limitation"]),
         cells(s["record_id"], ["direction"]), cells(s["record_id"], ["context"])]
        for s in snapshot["handoff"]["sources"]
    ]))
    gaps = {g["id"]: g for g in evidence["gaps"]}
    save("opportunity-matrix.svg", table_svg("Opportunity matrix — feasibility is Hypatia's assessment", caption, [
        {"label": "Opportunity", "width": 125}, {"label": "Gap / status", "width": 290},
        {"label": "Effort", "width": 110}, {"label": "Unknowns", "width": 325},
        {"label": "Author-proposal evidence", "width": 250},
    ], [
        [o["id"], f"{o['gap_id']}: {gaps[o['gap_id']]['status']}", o["feasibility"]["effort"],
         "; ".join(o["feasibility"]["unknowns"]) or "None recorded", ", ".join(o["direction_claim_ids"])]
        for o in evidence["opportunities"]
    ]))
    directions = []
    for opportunity in evidence["opportunities"]:
        gap = gaps[opportunity["gap_id"]]
        directions.append([
            f"{gap['id']}: {gap['statement']} [{', '.join(gap['claim_ids'])}] ({gap['status']})", "→",
            f"{opportunity['id']}: {opportunity_text(evidence, opportunity)} [{', '.join(opportunity['direction_claim_ids'])}]",
        ])
    save("gap-directions.svg", table_svg("Literature gaps → author-proposed directions", caption, [
        {"label": "Gap / source claims", "width": 475}, {"label": "Link", "width": 70},
        {"label": "Author direction / source claims", "width": 555},
    ], directions))
    save("review-flow.svg", svg("Callimachus review flow", caption, [
        f"{len(snapshot['ledger']['queries'])} recorded query audit entries → {len(snapshot['ledger']['records'])} deduplicated publications → {len(snapshot['handoff']['included_ids'])} final included publications.",
        f"Deferred: {sum(r['status'] == 'deferred' for r in snapshot['ledger']['records'])}.",
        f"Included with full text: {sum(s['kind'] == 'fulltext' for s in snapshot['handoff']['sources'])}; abstract-only: {sum(s['kind'] == 'abstract' for s in snapshot['handoff']['sources'])}.",
        "All final included sources have human curation dispositions. Access limitations are retained.",
    ]))
    save("evidence.json", json_text(evidence))
    save("context.json", json_text(context))
    for name in ("references.bib", "references.csv"):
        save(name, inside(snapshot["directory"], name).read_bytes())
    save("provenance.json", json_text({
        "handoff": snapshot["handoff"], "evidence_digest": evidence_hash, "renderer_version": renderer_version,
    }))
    load_snapshot(snapshot["root"], snapshot["handoff"]["revision"])
    if evidence_hash != evidence_digest(snapshot):
        raise ValueError("Evidence changed during rendering")
    staging.rename(directory)
    atomic_write(inside(root, "delivery.json"), delivery)
    return str(directory)

File: extensions/hypathia/render.ts
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { randomUUID } from "node:crypto";
import { join } from "node:path";
import { evidenceDigest, evidenceDirectory, loadContext, loadEvidence, validateDelivery } from "./evidence.js";
import { loadSnapshot, type Snapshot } from "./handoff.js";
import { atomicWrite, inside, json } from "./storage.js";
import type { Evidence, ResearchContext } from "./schema.js";

const md = (value: string) => value.replace(/[\\`*_{}[\]()<>#|!]/g, "\\$&").replace(/\r?\n/g, " ");
const xml = (value: string) => value.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
const cites = (ids: string[]) => ids.map(id => `[${id}](#claim-${id})`).join(", ");

export function shortlist(evidence: Evidence) {
  return evidence.opportunities.filter(o => o.feasibility.effort === "low" &&
    ["open-in-reviewed-corpus", "partly-addressed"].includes(evidence.gaps.find(g => g.id === o.gap_id)!.status));
}

function opportunityText(evidence: Evidence, opportunity: Evidence["opportunities"][number]): string {
  return opportunity.direction_claim_ids.map(id => evidence.claims.find(c => c.id === id)!.statement).join("; ");
}

export function report(snapshot: Snapshot, evidence: Evidence, context: ResearchContext): string {
  const { handoff, ledger } = snapshot;
  const lines = [
    `# Hypathia — ${md(ledger.question)}`, "",
    `Audience: ${md(context.audience)}. Search cutoff: **${handoff.cutoff}**.`,
    `Callimachus revision: \`${handoff.revision}\`. Final included publications: **${handoff.included_ids.length}**.`,
    "Publication counts are not independent-study counts or evidence-quality scores.", "",
    "## Executive summary", "",
    ...evidence.findings.slice(0, 5).map(f => `- ${md(f.statement)} (${cites(f.claim_ids)})`),
    evidence.findings.length ? "" : "No supported findings were established.",
    `${shortlist(evidence).length} direction(s) meet the stated low-effort criteria. Feasibility is Hypathia's assessment, not an author claim.`,
    "", "## Scope and methods", "",
    `Include: ${ledger.criteria.include.map(md).join("; ") || "None specified"}.`,
    `Exclude: ${ledger.criteria.exclude.map(md).join("; ") || "None specified"}.`,
    `Callimachus recorded ${ledger.queries.length} query audit entries and ${ledger.records.length} deduplicated records. All abstract screening and required human curation dispositions are complete.`,
    "The complete search audit and screening decisions are in the pinned Callimachus ledger. No additional search was performed by Hypathia.",
    "", "## Main findings and disagreements", "",
  ];
  for (const finding of evidence.findings) lines.push(
    `### ${finding.id}: ${md(finding.statement)}`, "",
    `Evidence: ${cites(finding.claim_ids)}. Appraisal: ${md(finding.strength)}.`,
    `Disagreement / counterevidence: ${finding.disagreements.length ? cites(finding.disagreements) : "none recorded; this does not establish consensus"}.`, "",
  );
  lines.push("## Literature-identified gaps", "");
  for (const gap of evidence.gaps) lines.push(
    `### ${gap.id}: ${md(gap.statement)}`, "",
    `Status: **${gap.status}**. Evidence: ${cites(gap.claim_ids)}.`,
    `Currency assessment: ${md(gap.currency.rationale)} (${gap.currency.checked_source_ids.length}/${handoff.included_ids.length} included publications checked).`,
    `Subsequent evidence: ${cites(gap.currency.claim_ids) || "none recorded"}.`, "",
  );
  lines.push("An open gap is scoped to this completed corpus and its cutoff; it is not proof of absence across the field.", "",
    "## Low-hanging fruit under stated constraints", "");
  if (!shortlist(evidence).length) lines.push("No directions can currently be shortlisted with adequate evidence and known resources.", "");
  for (const opportunity of shortlist(evidence)) lines.push(
    `- **${opportunity.id}** — ${md(opportunityText(evidence, opportunity))} (${cites(opportunity.direction_claim_ids)}); see the assessment below.`, "",
  );
  lines.push("## All author-proposed directions and feasibility", "");
  for (const opportunity of evidence.opportunities) lines.push(
    `### ${opportunity.id}: ${md(opportunityText(evidence, opportunity))}`, "",
    `Author-proposal evidence: ${cites(opportunity.direction_claim_ids)}. Gap: ${opportunity.gap_id}.`,
    `Hypathia effort assessment: **${opportunity.feasibility.effort}** — ${md(opportunity.feasibility.rationale)}.`,
    `Prerequisites: ${opportunity.feasibility.prerequisites.map(md).join("; ") || "unknown"}.`,
    `Unknowns: ${opportunity.feasibility.unknowns.map(md).join("; ") || "none recorded"}.`,
    `Researcher resource references: ${opportunity.feasibility.resource_ids.join(", ") || "none; feasibility remains unestablished"}.`, "",
  );
  lines.push("## Researcher constraints", "",
    ...context.resources.map(r => `- ${r.id}: ${md(r.description)}`),
    context.resources.length ? "" : "No resources supplied; low-effort shortlisting is disabled.", "",
    "## Coverage, access, and unresolved evidence", "");
  for (const source of handoff.sources) lines.push(
    `- ${md(source.record_id)}: ${source.kind}${source.limitation ? ` — ${md(source.limitation)}` : ""}. ${source.warnings.map(md).join(" ")}`,
  );
  for (const review of evidence.source_reviews) lines.push(`- ${md(review.source_id)} — ${review.status}: ${md(review.note)}`);
  for (const claim of evidence.claims.filter(c => c.verification.status !== "supported"))
    lines.push(`- ${claim.id}: **${claim.verification.status}** — ${md(claim.verification.reason)}.`);
  lines.push("", "## References", "");
  for (const record of ledger.records.filter(r => handoff.included_ids.includes(r.id)))
    lines.push(`- **${md(record.id)}**: ${md(record.authors.join(", "))}. ${md(record.title)} (${record.year ?? "year unknown"}). DOI: ${md(record.doi ?? "unavailable")}.`);
  lines.push("", "## Claim-to-source audit", "",
    "Exact quotation matching checks provenance. Semantic support and scientific strength remain explicit, reviewable assessments.", "");
  for (const claim of evidence.claims) {
    const source = handoff.sources.find(s => s.record_id === claim.source_id)!;
    lines.push(
      `<a id="claim-${claim.id}"></a>`, `### ${claim.id} — ${claim.kind}`, "",
      `${md(claim.statement)}`, "",
      `> ${md(claim.quote)}`, "",
      `Source: ${md(claim.source_id)}, ${source.kind === "abstract" ? "abstract" : `PDF page ${claim.page}`}; artifact \`${source.artifact}\`.`,
      `Verification: **${claim.verification.status}** — ${md(claim.verification.reason)}.`,
      `Appraisal: ${md(claim.appraisal)}. Independent-study identity: ${md(claim.study_id ?? "unknown")}.`, "",
    );
  }
  return lines.join("\n") + "\n";
}

function wrap(value: string, width = 110): string[] {
  return value.match(new RegExp(`.{1,${width}}(?:\\s|$)|.{1,${width}}`, "g"))?.map(s => s.trim()) ?? [""];
}

export function svg(title: string, description: string, rows: string[]): string {
  const lines = [title, description, ...rows.flatMap(row => [...wrap(row), ""])];
  const content = lines.map((line, i) =>
    `<text x="24" y="${38 + i * 25}" font-size="${i === 0 ? 22 : 15}">${xml(line)}</text>`).join("\n");
  return `<svg xmlns="http://www.w3.org/2000/svg" role="img" aria-labelledby="title description" viewBox="0 0 1150 ${70 + lines.length * 25}">
<title id="title">${xml(title)}</title><desc id="description">${xml(description)}</desc>
<rect width="100%" height="100%" fill="white"/><g fill="#172b4d" font-family="sans-serif">${content}</g></svg>\n`;
}

export function render(snapshot: Snapshot): string {
  loadSnapshot(snapshot.root, snapshot.handoff.revision);
  const evidence = loadEvidence(snapshot);
  const context = loadContext(snapshot);
  validateDelivery(snapshot, evidence);
  const digest = evidenceDigest(snapshot);
  const root = evidenceDirectory(snapshot);
  const directory = inside(root, `exports/${digest}`);
  if (existsSync(directory)) return directory;
  const staging = inside(root, `exports/${digest}.${randomUUID()}.partial`);
  mkdirSync(staging, { recursive: true });
  const save = (name: string, data: string | Buffer) => writeFileSync(join(staging, name), data, { flag: "wx" });
  save("report.md", report(snapshot, evidence, context));
  save("brief.md", [
    `# ${md(snapshot.ledger.question)}`, `Search cutoff: ${snapshot.handoff.cutoff}; revision: ${snapshot.handoff.revision}.`,
    "", "## Findings", ...evidence.findings.map(f => `- ${md(f.statement)} (${f.claim_ids.join(", ")}; see report audit). Strength: ${md(f.strength)}. Counterevidence: ${f.disagreements.join(", ") || "none recorded"}.`),
    "", "## Gaps", ...evidence.gaps.map(g => `- ${md(g.statement)} — ${g.status} (${g.claim_ids.join(", ")}).`),
    "", "## Feasible author-proposed directions",
    ...shortlist(evidence).map(o => `- ${md(opportunityText(evidence, o))} (${o.direction_claim_ids.join(", ")}). Hypathia feasibility assessment: ${md(o.feasibility.rationale)}.`),
    shortlist(evidence).length ? "" : "No adequately supported low-effort shortlist.",
    "", "## Limitations", "This summary inherits the report's corpus cutoff, access limitations, uncertainty, and researcher constraints.",
    ...snapshot.handoff.sources.filter(s => s.limitation).map(s => `- ${md(s.record_id)}: ${md(s.limitation)}`),
    ...evidence.claims.filter(c => c.verification.status !== "supported").map(c => `- ${c.id}: ${c.verification.status} — ${md(c.verification.reason)}`),
  ].join("\n") + "\n");
  const caption = `Cutoff ${snapshot.handoff.cutoff}. ${snapshot.handoff.included_ids.length} included publications; counts are not evidence strength.`;
  save("evidence-matrix.svg", svg("Evidence matrix", caption, [
    "Legend: supported / uncertain / contradicted are verification decisions. IDs resolve in the report audit.",
    ...snapshot.handoff.sources.map(s => `${s.record_id} [${s.kind}]: ${evidence.claims.filter(c => c.source_id === s.record_id).map(c => `${c.id} ${c.kind} (${c.verification.status})`).join("; ") || "no claims extracted"}`),
  ]));
  save("opportunity-matrix.svg", svg("Opportunity matrix", caption, [
    "Columns: opportunity | gap status | effort assessment | unknowns | author-proposal evidence",
    ...evidence.opportunities.map(o => `${o.id} | ${evidence.gaps.find(g => g.id === o.gap_id)!.status} | ${o.feasibility.effort} | ${o.feasibility.unknowns.join("; ") || "none recorded"} | ${o.direction_claim_ids.join(", ")}`),
  ]));
  save("gap-directions.svg", svg("Literature gaps → author-proposed directions", caption,
    evidence.opportunities.map(o => `${o.gap_id} (${evidence.gaps.find(g => g.id === o.gap_id)!.claim_ids.join(", ")}) → ${o.id}: ${opportunityText(evidence, o)} (${o.direction_claim_ids.join(", ")})`)));
  save("review-flow.svg", svg("Callimachus review flow", caption, [
    `${snapshot.ledger.queries.length} recorded query audit entries → ${snapshot.ledger.records.length} deduplicated publications → ${snapshot.handoff.included_ids.length} final included publications.`,
    `Deferred: ${snapshot.ledger.records.filter(r => r.status === "deferred").length}.`,
    `Included with full text: ${snapshot.handoff.sources.filter(s => s.kind === "fulltext").length}; abstract-only: ${snapshot.handoff.sources.filter(s => s.kind === "abstract").length}.`,
    "All final included sources have human curation dispositions. Access limitations are retained.",
  ]));
  save("evidence.json", json(evidence));
  save("context.json", json(context));
  for (const name of ["references.bib", "references.csv"]) save(name, readFileSync(inside(snapshot.directory, name)));
  save("provenance.json", json({ handoff: snapshot.handoff, evidence_digest: digest }));
  loadSnapshot(snapshot.root, snapshot.handoff.revision);
  if (digest !== evidenceDigest(snapshot)) throw new Error("Evidence changed during rendering");
  renameSync(staging, directory);
  atomicWrite(inside(root, "delivery.json"), json({ digest, directory }));
  return directory;
}
